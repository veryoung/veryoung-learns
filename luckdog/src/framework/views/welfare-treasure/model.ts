import { addKeylink, KeylinkCmd, logError } from '@/luckdog';
import { showToast } from '@/luckbox';
import { doFetch } from '@/luckbox/fetch';

import { LOG_PREFIX, TreasureStatus, POWER_COUNTDOWN } from './constant';
import { TreasureContext, TreasureState, TreasureChestStatus, GetTreasureChestReply, OpenTreasureChestReply, ReceiveVideoAdRewardReply } from './types';
import { isSameDay, refreshToken, getCountdown } from './utils';
import reporter from './reporter';

const TAG = `${LOG_PREFIX}-model`;

// token 刷新标记位
let isTokenRefreshed = false;

// 获取 tprc 错误码
const getTrpcErrorCode = (result) => {
  const header = result?.headers?.['Trpc-Func-Ret'] || result?.headers?.['trpc-func-ret'];
  const value = Array.isArray(header) ? header[0] : header;
  return value || '';
};

/**
 * token 尝试刷新
 */
const tryToRefreshToken = async () => {
  // 尝试刷新一次 token 再重新执行
  if (!isTokenRefreshed) {
    isTokenRefreshed = true;
    await refreshToken();
    return true;
  }

  return false;
};

/**
 * 查询用户的福袋状态
 */
export const fetchTreasureStatus = async (context: TreasureContext, isLogin = true): Promise<TreasureStatus> => {
  const { error, result } = await doFetch<GetTreasureChestReply>(
    '/trpc.pcg_novel.welfare_incentives.WelfareIncentives/GetTreasureChest',
    {
      method: 'POST',
      headers: {
        'Content-Type': 'application/json',
      },
      body: JSON.stringify({ from: 1 }),
    },
  );

  if (result?.body?.ret_code === 0) {
    const status = resolveTreasureStatus(context, result.body);

    // floating 曝光上报
    reporter.reportFloatingExposure({
      statusType: reporter.getStatusType(status, isLogin),
    });

    return status;
  }

  if (!result) {
    // 链路错误，如网络异常
    logError(`fetch status failed [1]: ${error}`, TAG);
  } else if (!result.body) {
    // 服务错误，如鉴权失败
    const errcode = getTrpcErrorCode(result);

    // 默认当 token 过期处理，尝试刷新一次 token
    const isRefreshed = await tryToRefreshToken();
    if (isRefreshed) return fetchTreasureStatus(context);

    logError(`fetch status failed [2]: ${errcode || -1}`, TAG);
  } else {
    // 业务错误，如参数有误
    logError(`fetch status failed [3]: ${result.body.ret_code || -999}`, TAG);
  }


  return TreasureStatus.UNKNOWN;
};

/**
 * 处理用户福袋状态查询结果
 */
const resolveTreasureStatus = (
  { setState }: TreasureContext,
  data: { status: TreasureChestStatus },
): TreasureStatus => {
  const { status } = data;
  const {
    can_open: canOpen,
    total_times: totalTimes,
    completed_times: completedTimes,
    svr_ts: svrTs,
    next_ts: nextTs,
  } = status;
  const { UNKNOWN, ENABLED, COOLING, DISABLED, NOTASK } = TreasureStatus;

  let state: Partial<TreasureState> = {
    isUnfreeze: false,
  };

  if (canOpen) {
    // 可以开启福袋
    state = { ...state, status: ENABLED };
  } else if (totalTimes === 0) {
    // 用户没有福袋任务
    state = { ...state, status: NOTASK };
  } else if (totalTimes === completedTimes) {
    // 已完成次数等于总次数，表示任务全部完成
    state = { ...state, status: DISABLED };
  } else if (nextTs > 0) {
    // 冷却中
    const isTomorrow = !isSameDay(svrTs, svrTs + nextTs);

    // 临界：冷却时间超过当天最晚时间，任务不可完成
    if (isTomorrow) {
      state = { ...state, status: DISABLED };
    } else {
      state = { ...state, status: COOLING, coolingTime: nextTs };
    }
  }

  addKeylink(`福袋状态: ${JSON.stringify({ status })}`, TAG);
  addKeylink(`welfare-floating-status=${TreasureStatus?.[state.status || '']}`, KeylinkCmd.PR_INFO_SUM);

  setState(state);

  return state.status || UNKNOWN;
};

/**
 * 开启福袋
 */
export const openTreasure = async (context: TreasureContext) => {
  const { error, result } = await doFetch<OpenTreasureChestReply>(
    '/trpc.pcg_novel.welfare_incentives.WelfareIncentivesHTTP/OpenTreasureChest',
    {
      method: 'POST',
      headers: {
        'Content-Type': 'application/json',
      },
      body: JSON.stringify({ from: 1 }),
    },
  );

  if (result?.body?.ret_code === 0) {
    // 数据正常
    return resolveOpenTreasure(context, result.body);
  }

  if (!result) {
    // 链路错误，如网络异常
    logError(`open treasure failed [1]: ${error}`, TAG);
  } else if (!result.body) {
    // 服务错误，如鉴权失败
    const errcode = getTrpcErrorCode(result);

    // 默认当 token 过期处理，尝试刷新一次 token
    const isRefreshed = await tryToRefreshToken();
    if (isRefreshed) return openTreasure(context);

    logError(`open treasure failed [2]: ${errcode || -1}`, TAG);
  } else {
    // 业务错误，如参数有误
    logError(`open treasure failed [3]: ${result.body.ret_code || -999}`, TAG);
  }

  showToast('活动太火爆了，请稍后再试');
};

/**
 * 处理用户福袋开启结果
 */
const resolveOpenTreasure = (context: TreasureContext, data: OpenTreasureChestReply) => {
  const { setState } = context;
  const { result, show_video_ad: isShowVideoAd, video_ad_max_gold: videoAdMax } = data;

  // 展示弹窗
  setState({
    isWeakModalShow: true,
    rewardNum: result.reward_num,
    powerRewardMax: (isShowVideoAd && videoAdMax) || 0,
  });
  // 弹窗曝光上报
  reporter.reportWeakModalExposure();

  // 更新挂件状态
  resolveTreasureStatus(context, data);
};

/**
 * 领取激励视频奖励
 */
export const receiveVideoAdReward = async (context: TreasureContext) => {
  const { error, result } = await doFetch<ReceiveVideoAdRewardReply>(
    '/trpc.pcg_novel.welfare_incentives.WelfareIncentivesHTTP/ReceiveVideoAdReward',
    {
      method: 'POST',
      headers: {
        'Content-Type': 'application/json',
      },
      body: JSON.stringify({ business_id: 314 }),
    },
  );

  if (result?.body?.ret_code === 0) {
    // 数据正常
    return resolveReceiveVideoAdReward(context, result.body);
  }

  if (!result) {
    // 链路错误，如网络异常
    logError(`receive video ad award failed [1]: ${error}`, TAG);
  } else if (!result.body) {
    // 服务错误，如鉴权失败
    const errcode = getTrpcErrorCode(result);

    // 默认当 token 过期处理，尝试刷新一次 token
    const isRefreshed = await tryToRefreshToken();
    if (isRefreshed) return receiveVideoAdReward(context);

    logError(`receive video ad award failed [2]: ${errcode || -1}`, TAG);
  } else {
    // 业务错误，如参数有误
    logError(`receive video ad award failed [3]: ${result.body.ret_code || -999}`, TAG);
  }

  showToast('活动太火爆了，请稍后再试');
};

/**
 * 处理用户福袋开启结果
 */
const resolveReceiveVideoAdReward = (context: TreasureContext, data: ReceiveVideoAdRewardReply) => {
  const { setState } = context;
  const { reward_num: rewardNum } = data;

  // 展示弹窗
  setState({
    isPowerModalShow: true,
    powerRewardNum: rewardNum,
    powerModalCountdown: POWER_COUNTDOWN,
  });

  // 开启关闭倒计时
  getCountdown().set(POWER_COUNTDOWN, (remainTime: number) => {
    let state: Partial<TreasureState> = {
      powerModalCountdown: remainTime,
    };

    if (remainTime === 0) {
      state = {
        ...state,
        isPowerModalShow: false,
      };
      // 清除倒计时
      getCountdown().clear();
    }

    setState(state);
  });

  // 弹窗曝光上报
  reporter.reportPowerModalExposure();
};
