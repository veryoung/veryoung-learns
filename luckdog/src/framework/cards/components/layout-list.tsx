import React from 'react';
import { StyleSheet, View } from '@tencent/hippy-react-qb';

import FeedsProtect from '../../../mixins/FeedsProtect';
import { CardLayout } from '../../protocol/card';

interface CompProps<T> {
  /** 最外层容器的布局方式 */
  isRow: boolean;
  /** 行列组合后的item布局列表 */
  layoutList: CardLayout<T>[];
  /** item的行间距 */
  rowSpacing?: number;
  /** item的列间距 */
  columnSpacing?: number;
  /** item的渲染函数 */
  onItemRender: (
    /** 是否是行列结合布局后的小item */
    isSmallItem: boolean,
    /** layout索引 */
    layoutIdx: number,
    /** dataItem索引 */
    itemIdx: number,
  ) => JSX.Element | null;
  /** item的layout回调 */
  onItemLayout?: (
    /** 事件回调参数，主要取layout属性 */
    event: any,
    /** 是否是行列结合布局后的小item */
    isSmallItem: boolean,
    /** layout索引 */
    layoutIdx: number,
    /** dataItem索引 */
    itemIdx: number,
    /** 扩展信息，一般传dataItem对象 */
    extInfo?: Record<string, any>,
  ) => void;
}

/** 判断当前dataList所在layout是否处于行列结合后的小卡 */
const isInCardSmallItem = <T extends unknown>(
  dataList: T[],
  layoutList: CardLayout<T>[],
): boolean => layoutList.some(item => dataList.length > item.dataList.length);


/**
 * 行列各种组合的布局参考:
 * https://codesign.woa.com/s/YDgGjYrrNb9wEVQ
 * https://codesign.woa.com/s/XgRxnjPM8E9Lmqr
 */
@FeedsProtect.protect
export class LayoutList<T> extends React.Component<CompProps<T>> {
  public constructor(props: CompProps<T>) {
    super(props);
  }

  public render() {
    const { isRow, layoutList = [], rowSpacing = 0, columnSpacing = 0 } = this.props;
    // 外侧容器样式
    const contentStyle = isRow ? styles.row : styles.column;
    // 每个layout间距
    const spacingStyle = isRow ? { marginLeft: columnSpacing } : { marginTop: rowSpacing };
    return <View collapsable={false} style={contentStyle}>
      {
        layoutList.map((layoutItem, layoutIdx) => <View
          key={`layoutItem_${layoutIdx}`}
          style={layoutIdx > 0 ? spacingStyle : {}}
        >
          {this.renderLayout(layoutItem, layoutIdx)}
        </View>)
      }
    </View>;
  }

  /** 外层layout渲染 */
  private renderLayout = (layoutItem: CardLayout<T>, layoutIdx: number) => {
    const { layoutList = [], rowSpacing = 0, columnSpacing = 0, onItemRender, onItemLayout } = this.props;
    const { isRow, dataList = [] } = layoutItem || {};
    // layout容器样式
    const layoutStyle = isRow ? styles.row : styles.column;
    // 每个item间距
    const spacingStyle = isRow ? { marginLeft: columnSpacing } : { marginTop: rowSpacing };
    const isSmallItem = isInCardSmallItem(dataList, layoutList);
    return <View style={layoutStyle} collapsable={false}>
      {
        dataList.map((_dataItem, itemIdx) => <View
          key={`listItem_${itemIdx}`}
          style={itemIdx > 0 ? spacingStyle : {}}
          onLayout={event => onItemLayout?.(event, isSmallItem, layoutIdx, itemIdx)}
        >
          {onItemRender(isSmallItem, layoutIdx, itemIdx)}
        </View>)
      }
    </View>;
  };
}

const styles = StyleSheet.create({
  row: {
    flexDirection: 'row',
  },
  column: {
    flexDirection: 'column',
  },
});
