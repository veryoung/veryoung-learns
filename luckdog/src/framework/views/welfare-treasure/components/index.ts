export { Floating } from './floating';
export { WeakModal } from './weak-modal';
export { PowerModal } from './power-modal';
