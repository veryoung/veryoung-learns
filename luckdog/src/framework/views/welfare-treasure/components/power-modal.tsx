import React from 'react';
import { View, Image, Text, Modal, StyleSheet } from '@tencent/hippy-react-qb';

import { getWidthHeight, isIOS } from '@/framework/utils/device';

import { TreasureContext } from '../types';
import source from '@/framework/FeedsIcon';
import { getDeviceVisitor } from '@/luckbox';
import FeedsTheme from '@/framework/FeedsTheme';
import { SkinModelType } from '@/entity';
import FeedsConst from '@/framework/FeedsConst';

interface Props {
  context: TreasureContext;
  onClick: (context: TreasureContext, isConfirm: boolean) => any;
}

const IMAGES = source.welfareTreasure;

const Z_INDEX_1 = 1;
const Z_INDEX_2 = 2;
const Z_INDEX_3 = 3;
const Z_INDEX_4 = 4;
const WELFARE_PACKET_BTN_BG = ['rgba(255,222,165,1)', isIOS ? 'rgba(255,222,165,1)' : 'rgba(172,144,102,1)'];
const WELFARE_PACKET_BTN_TEXT = ['rgba(97,54,16,1)', 'rgba(97,54,16,1)'];
const WELFARE_AWARD_LABEL_BG = ['rgba(232,174,112,1)', isIOS ? 'rgba(232,174,112,1)' : 'rgba(172,144,102,1)'];
const WELFARE_AWARD_LABEL_TEXT = ['rgba(97,54,16,1)', isIOS ? 'rgba(97,54,16,1)' : 'rgba(97,54,16,0.5)'];
const WELFARE_AWARD_NUM_TEXT = ['rgba(254,222,163,1)', isIOS ? 'rgba(254,222,163,1)' : 'rgba(254,222,163,0.5)'];
const WELFARE_AWARD_LABEL_BORDER = ['rgba(255,255,255,1)', isIOS ? 'rgba(255,255,255,1)' : 'rgba(97,54,16,1)'];
const isAndroid = getDeviceVisitor().isAdr();

export const PowerModal = (props: Props) => {
  const { context, onClick } = props;
  const { state: { isPowerModalShow, powerRewardNum, powerModalCountdown } } = context;

  if (!isPowerModalShow || !FeedsConst.getGlobalConfKV('initActive')) return null;

  return (
    <Modal
      visible
      transparent
      animationType="fade"
      onRequestClose={() => onClick(context, false)}
      supportedOrientations={['portrait']}
    >
      <View style={styles.mask}>
        {FeedsTheme.returnSkinId() === SkinModelType.NIGHT && isIOS && <View style={styles.wrapperMask}></View>}
        <View style={styles.wrapper}>
          {/* 光效 */}
          <Image
            style={styles.lightBg}
            source={IMAGES.light}
            noPicMode={{ enable: false }}
            nightMode={{ enable: isAndroid }}
          />
          {/* 奖励 */}
          <View style={styles.awardWrap}>
            {/* 架子背景 */}
            <Image
              style={styles.awardBg}
              source={IMAGES.awardBg}
              noPicMode={{ enable: false }}
              nightMode={{ enable: isAndroid }}
            />
            <Text style={styles.welfareNum}>x1</Text>
            {/* 奖励列表 */}
            <View style={styles.awardList}>
              <View style={styles.awardItem}>
                <Image
                  style={[styles.awardIcon, styles.awardIconLg]}
                  source={IMAGES.coin}
                  noPicMode={{ enable: false }}
                  nightMode={{ enable: isAndroid }}
                />
                <Image
                  style={[styles.shadowIcon, styles.shadowIconLg]}
                  source={IMAGES.shadow}
                  noPicMode={{ enable: false }}
                  nightMode={{ enable: isAndroid }}
                />
                <View style={[styles.labelWrap, styles.labelWrapLg]}>
                  <Text style={[styles.labelText, styles.labelTextLg]}>{powerRewardNum}金币</Text>
                </View>
              </View>
            </View>
          </View>
          {/* 知道了 按钮 */}
          <View style={styles.btnWrap}>
            <Text style={styles.btnText} onClick={() => onClick(context, true)}>我知道了({powerModalCountdown}秒)</Text>
          </View>
          {/* 关闭 icon */}
          <Image
            style={styles.closeIcon}
            source={IMAGES.modalClose}
            onClick={() => onClick(context, false)}
            noPicMode={{ enable: false }}
            nightMode={{ enable: isAndroid }}
          />
        </View>
      </View>
    </Modal>
  );
};

// css
const styles = StyleSheet.create({
  mask: {
    width: getWidthHeight().width,
    height: getWidthHeight().height,
    flex: 1,
    alignItems: 'center',
    justifyContent: 'center',
    backgroundColor: 'rgba(0, 0, 0, 0.4)',
  },
  wrapper: {
    width: getWidthHeight().width,
    flex: 1,
    position: 'relative',
    alignItems: 'center',
    justifyContent: 'center',
  },
  wrapperMask: {
    width: getWidthHeight().width,
    height: getWidthHeight().height,
    position: 'absolute',
    backgroundColor: 'rgba(0, 0, 0, 0.4)',
    zIndex: Z_INDEX_4,
  },
  lightBg: {
    position: 'absolute',
    width: getWidthHeight().width,
    height: getWidthHeight().width,
  },
  awardWrap: {
    alignItems: 'center',
    justifyContent: 'flex-end',
  },
  awardBg: {
    width: (660 / 750) * getWidthHeight().width,
    height: (510 / 750) * getWidthHeight().width,
  },
  awardList: {
    position: 'absolute',
    bottom: 46,
    flexDirection: 'row',
    justifyContent: 'center',
    alignItems: 'center',
  },
  awardItem: {
    position: 'relative',
    alignItems: 'center',
    justifyContent: 'flex-end',
  },
  awardIcon: {
    position: 'relative',
    zIndex: Z_INDEX_2,
  },
  awardIconLg: {
    width: 140,
    height: 140,
  },
  shadowIcon: {
    position: 'absolute',
    zIndex: Z_INDEX_1,
    bottom: -3,
  },
  shadowIconLg: {
    width: 120,
    height: 36,
  },
  btnWrap: {
    height: 44,
    borderRadius: 44,
    alignItems: 'center',
    justifyContent: 'center',
    backgroundColors: WELFARE_PACKET_BTN_BG,
    marginBottom: 31,
    marginTop: -30,
    paddingHorizontal: 24,
  },
  btnText: {
    colors: WELFARE_PACKET_BTN_TEXT,
    fontSize: 17,
  },
  closeIcon: {
    width: 30,
    height: 30,
  },
  labelWrap: {
    position: 'absolute',
    backgroundColors: WELFARE_AWARD_LABEL_BG,
    paddingLeft: 8,
    paddingRight: 8,
    borderWidth: 1,
    borderColors: WELFARE_AWARD_LABEL_BORDER,
    alignItems: 'center',
    justifyContent: 'center',
    zIndex: Z_INDEX_3,
  },
  labelWrapLg: {
    height: 20,
    borderRadius: 20,
    top: 10,
    right: 12,
  },
  labelText: {
    colors: WELFARE_AWARD_LABEL_TEXT,
    fontWeight: 'bold',
  },
  labelTextLg: {
    fontSize: 12,
  },
  welfareNum: {
    position: 'absolute',
    top: 18,
    left: 255,
    fontSize: 24,
    colors: WELFARE_AWARD_NUM_TEXT,
    zIndex: Z_INDEX_3,
  },
});
