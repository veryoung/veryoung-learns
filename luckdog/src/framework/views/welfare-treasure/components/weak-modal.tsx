import React from 'react';
import { View, Image, Text, Modal, StyleSheet } from '@tencent/hippy-react-qb';

import { getWidthHeight } from '@/framework/utils/device';

import { TreasureContext } from '../types';
import source from '@/framework/FeedsIcon';
import FeedsTheme from '@/framework/FeedsTheme';
import FeedsConst from '@/framework/FeedsConst';

interface Props {
  context: TreasureContext;
  onClick: (context: TreasureContext, isConfirm: boolean) => any;
}

const IMAGES = source.welfareTreasure;

export const WeakModal = (props: Props) => {
  const { context, onClick } = props;
  const { state: { isWeakModalShow, rewardNum, powerRewardMax } } = context;

  if (!isWeakModalShow || !FeedsConst.getGlobalConfKV('initActive')) return null;

  return (
    <Modal
      visible
      transparent
      animationType="fade"
      onRequestClose={() => onClick(context, false)}
      supportedOrientations={['portrait']}
    >
      <View style={styles.mask}>
        <View style={styles.wrapper}>
          <View style={styles.modal}>
            <Text style={styles.title}>恭喜你</Text>
            <View style={styles.tip}>
              <Text style={styles.label}>获得</Text>
              <Text style={styles.strong}>{rewardNum}</Text>
              <Text style={styles.label}>金币</Text>
            </View>
            <Image
              style={styles.image}
              source={IMAGES.modalImg}
              noPicMode={{ enable: false }}
            />
            <View style={styles.button}>
              <Image
                style={styles.buttonBg}
                source={IMAGES.modalBtn}
                noPicMode={{ enable: false }}
              />
              <Text style={styles.buttonText} onClick={() => onClick(context, true)}>
                {powerRewardMax ? `看视频再领最高${powerRewardMax}金币` : '稍后再来'}
              </Text>
            </View>
          </View>
          <Image
            style={styles.close}
            source={IMAGES.modalClose}
            onClick={() => onClick(context, false)}
            noPicMode={{ enable: false }}
            // nightMode={{ enable: isAndroid }}
          />
        </View>
      </View>
    </Modal>
  );
};

// css
const styles = StyleSheet.create({
  mask: {
    width: getWidthHeight().width,
    height: getWidthHeight().height,
    flex: 1,
    alignItems: 'center',
    justifyContent: 'center',
    backgroundColor: 'rgba(0, 0, 0, 0.6)',
  },
  wrapper: {
    width: getWidthHeight().width,
    flex: 1,
    position: 'relative',
    alignItems: 'center',
    justifyContent: 'center',
  },
  modal: {
    alignItems: 'center',
    width: 266,
    height: 294.5,
    backgroundColors: FeedsTheme.SkinColor.A5,
    borderRadius: 16,
  },
  title: {
    height: 24,
    marginTop: 28,
    fontFamily: 'PingFangSC-Medium',
    fontSize: 20,
    colors: FeedsTheme.SkinColor.B5,
    textAlign: 'center',
    lineHeight: 24,
  },
  tip: {
    height: 21,
    marginTop: 4,
    flexDirection: 'row',
    alignItems: 'center',
    justifyContent: 'center',
  },
  label: {
    fontFamily: 'PingFangSC-Regular',
    fontSize: 14,
    colors: FeedsTheme.SkinColor.A2,
  },
  strong: {
    fontFamily: 'PingFangSC-Regular',
    fontSize: 14,
    colors: FeedsTheme.SkinColor.B5,
  },
  image: {
    width: 176,
    height: 104,
    marginTop: 26,
  },
  button: {
    position: 'relative',
    width: 228,
    height: 44,
    marginTop: 16,
  },
  buttonBg: {
    position: 'absolute',
    left: 0,
    top: 0,
    width: 228,
    height: 44,
  },
  buttonText: {
    position: 'absolute',
    left: 0,
    top: 0,
    width: 228,
    height: 44,
    lineHeight: 44,
    fontFamily: 'PingFangSC-Medium',
    fontSize: 17,
    color: '#613610',
    textAlign: 'center',
  },
  close: {
    width: 30,
    height: 30,
    opacity: 0.8,
    marginTop: 28,
  },
});
