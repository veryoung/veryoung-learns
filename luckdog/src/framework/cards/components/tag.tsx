/* eslint-disable no-loop-func */
import React from 'react';
import {
  View,
  Text,
  ScrollView,
  StyleSheet,
} from '@tencent/hippy-react-qb';
import FeedsProtect from '../../../mixins/FeedsProtect';
import { shouldComponentUpdate, countReRender } from '@tencent/luckbox-react-optimize';
import { CommonProps } from '../../../entity';
import { FeedsTheme } from '../../../feeds-styles/tab-22/components/utils';
import FeedsUtils from '../../FeedsUtils';
import { Category } from '../../protocol';

const tagsMargin = 8;
const tagsBottom = 8;

const tagsBgColors = [
  FeedsTheme.SkinColor.N2_1,
  FeedsTheme.SkinColor.N3_1,
  FeedsTheme.SkinColor.N5_1,
  FeedsTheme.SkinColor.N6_1,
  FeedsTheme.SkinColor.N7_1,
];
interface Props extends CommonProps {
  tags:  Category[];
  line: number;
  cardIndex: number;
}

/** 标签卡 */
@FeedsProtect.protect
export class Tags extends React.Component<Props> {
  public BookShelfScrollView: any;

  public constructor(props) {
    super(props);
    this.shouldComponentUpdate = shouldComponentUpdate(this, 'Tags');
  }

  public onScrollEndDrag = (e) => {
    const index = Math.floor(e.contentOffset.x / 160);
    this.props.doBeaconBySlide?.({ ext_data1: index });
  };

  public spliceGroup = (array, line) => {
    const length = Math.ceil(array.length / line);
    const newArray: any[] = [];
    for (let i = 1; i <= line; i++) {
      if (i !== line) {
        newArray.push(array.splice(0, length));
      } else {
        newArray.push(array.splice(0, array.length));
      }
    }
    return newArray;
  };

  public clickHandle = (tag, start) => {
    const {
      doBeaconByClick,
    } = this.props || {};
    doBeaconByClick?.({
      ext_data1: start - 1,
      ext_data2: tag.sTitle,
      bigdata_contentid: '',
    });
    FeedsUtils.doLoadUrl(tag?.jumpUrl);
  };

  public render() {
    countReRender(this, 'Tags');
    const {
      tags = [],
      line = 2,
    } = this.props || {};
    if (tags.length <= 0) {
      return null;
    }
    const tagsArray = this.spliceGroup([...tags], line);
    const tagsView: any[] = [];
    const bgColorBaseIndex = Math.floor(Math.random() * 5);
    for (let i = 0; i < tagsArray.length; i++) {
      const tagView: any[] = [];
      tagsArray[i].forEach((e, j) => {
        const bgColorIndex = (bgColorBaseIndex + i + j) % 5;
        const index = i > 0 ? (i * tagsArray[i - 1].length) + j + 1 : j + 1;
        tagView.push(<View
          collapsable={false}
          style={[
            styles.tag,
            { marginRight: j === tagsArray[i].length - 1 ? 0 : tagsMargin },
            { backgroundColors: tagsBgColors[bgColorIndex] },
          ]}
          key={`${e.name}-${index}`}
          onClick={() => this.clickHandle(e, index || 1)}
        >
          <Text style={styles.tagText}>{e.name}</Text>
        </View>);
      });
      tagsView.push(<View
        collapsable={false}
        style={[
          styles.tagWarp,
          { marginBottom: i === tagsArray.length - 1 ? 0 : tagsBottom },
        ]}
        key={`${i}`}
      >
        {tagView}
      </View>);
    }

    return (
      <ScrollView
        contentContainerStyle={styles.imageContainer}
        horizontal
        showsHorizontalScrollIndicator={false}
        sendMomentumEvents
        ref={(ref) => {
          this.BookShelfScrollView = ref;
        }}
        onScrollEndDrag={this.onScrollEndDrag}
        scrollEventThrottle={100}
      >
        <View
          style={[styles.tagsWarp]}
          collapsable={false}
        >
          {tagsView}
        </View>
      </ScrollView>
    );
  }
}

const styles = StyleSheet.create({
  imageContainer: {
    paddingBottom: 16,
    paddingHorizontal: 16,
  },
  tagsWarp: {
    flexDirection: 'column',
  },
  tagWarp: {
    flexDirection: 'row',
  },
  tag: {
    paddingLeft: 24,
    paddingRight: 24,
    paddingTop: 8,
    paddingBottom: 8,
    borderRadius: 18,
    flexDirection: 'row',
    alignItems: 'center',
    justifyContent: 'center',
  },
  tagText: {
    colors: FeedsTheme.SkinColor.N1,
  },
});
