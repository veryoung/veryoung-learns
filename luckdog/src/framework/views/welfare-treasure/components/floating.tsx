import React from 'react';
import { View, Image, Text, StyleSheet } from '@tencent/hippy-react-qb';

import { CountDown } from '@/components/count-down';
import { WelfarePendantShowType } from '@/presenters/welfare-floating-controller';

import { callbackWrap } from '../utils';
import { SHOWABLE_STATUS, TreasureStatus } from '../constant';
import { TreasureContext } from '../types';
import source from '@/framework/FeedsIcon';
import { getDeviceVisitor } from '@/luckbox';
import FeedsTheme from '@/framework/FeedsTheme';

interface Props {
  context: TreasureContext;
  onClick: (context: TreasureContext) => any;
  onClose: (context: TreasureContext) => any;
}

const IMAGES = source.welfareTreasure;
const isAndroid = getDeviceVisitor().isAdr();

const countDownStyles = {
  wrap: {
    position: 'absolute',
    left: 0,
    top: 0,
    zIndex: 2,
    width: 56,
    height: 20,
    flexDirection: 'row',
    justifyContent: 'center',
  },
  time: {
    height: 20,
    lineHeight: 18,
    fontSize: 10,
    textAlign: 'center',
    fontFamily: 'PingFangSC-Semibold',
    colors: FeedsTheme.SkinColor.A5,
  },
  spliter: {
    wrap: { width: 3, justifyContent: 'center', alignItems: 'center' },
    dot: { width: 1.5, height: 1.5, borderRadius: 1, backgroundColors: FeedsTheme.SkinColor.A5, marginTop: 2 },
  },
};

/**
 * 倒计时结束回调
 */
const onTimeEnd = async ({ setState }: TreasureContext) => {
  // 预加载解冻动效图片
  // await preloadImg(UNFREEZE_FRAME);

  setState({
    status: TreasureStatus.ENABLED,
    isUnfreeze: true,
  });
};

/**
 * 悬浮挂件提示文案
 */
const getFloatingTip = (context: TreasureContext) => {
  const { state } = context;
  const { status, coolingTime } = state;

  const strategies = {
    [TreasureStatus.ENABLED]: <Text style={styles.tip}>开福袋</Text>,
    [TreasureStatus.DISABLED]: <Text style={styles.tip}>明日再来</Text>,
    [TreasureStatus.COOLING]: (
      <CountDown
        key={coolingTime}
        remainTime={coolingTime}
        styles={countDownStyles}
        onTimeEnd={() => onTimeEnd(context)}
      />
    ),
  };

  return strategies[status];
};

/**
 * 悬浮挂件样式
 */
const getFloatingStyle = (context: TreasureContext) => {
  const styleList: Record<string, any>[] = [styles.floating];
  const { props: { showType } } = context;

  if (showType === WelfarePendantShowType.ASIDE) {
    styleList.push(styles.floatingAside);
  }

  if (showType === WelfarePendantShowType.HIDDEN) {
    styleList.push(styles.floatingHidden);
  }

  // iOS 夜间模式适配
  if (FeedsTheme.isNightMode() && !isAndroid) {
    styleList.push({ opacity: 0.7 });
  }

  return styleList;
};

export const Floating = (props: Props) => {
  const { context, onClick, onClose } = props;
  const { state: { status } } = context;

  // 非可展示状态时，不展示福袋
  if (!SHOWABLE_STATUS.includes(status)) return null;

  return (
    <View style={getFloatingStyle(context)} onClick={callbackWrap(() => onClick(context))}>
      <Image
        style={styles.icon}
        source={IMAGES.floating}
        noPicMode={{ enable: false }}
        nightMode={{ enable: isAndroid }}
      ></Image>
      <View style={styles.tipBox}>
        <Image
          style={styles.tipBg}
          source={IMAGES.floatingBtn}
          noPicMode={{ enable: false }}
          nightMode={{ enable: isAndroid }}
        ></Image>
        {getFloatingTip(context)}
      </View>
      <Image
        style={styles.close}
        source={IMAGES.floatingClose}
        onClick={callbackWrap(() => onClose(context))}
        noPicMode={{ enable: false }}
        nightMode={{ enable: isAndroid }}
      ></Image>
    </View>
  );
};

// css
const styles = StyleSheet.create({
  floating: {
    display: 'block',
    position: 'absolute',
    zIndex: 2,
    right: 13,
    bottom: 110,
    width: 90,
    height: 90,
  },
  floatingAside: {
    right: -51,
    opacity: 0.5,
  },
  floatingHidden: {
    display: 'none',
  },
  icon: {
    width: 90,
    height: 90,
  },
  tipBox: {
    position: 'absolute',
    left: 17,
    bottom: 9,
    width: 56,
    height: 20,
  },
  tipBg: {
    width: 56,
    height: 20,
  },
  tip: {
    position: 'absolute',
    left: 0,
    top: 0,
    zIndex: 2,
    width: 56,
    height: 20,
    lineHeight: 18,
    textAlign: 'center',
    fontFamily: 'PingFangSC-Semibold',
    fontSize: 10,
    colors: FeedsTheme.SkinColor.A5,
  },
  close: {
    position: 'absolute',
    right: 6,
    top: 6,
    width: 14,
    height: 14,
  },
});
