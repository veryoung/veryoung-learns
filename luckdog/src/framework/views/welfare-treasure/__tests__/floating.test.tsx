import React from 'react';
import renderer from 'react-test-renderer';

import { Floating } from '../components/floating';
import { getContext } from './model.test';

describe('render floating', () => {
  it('should render floating success', () => {
    const props = {
      context: getContext(),
      onClick: jest.fn(),
      onClose: jest.fn(),
    };
    const tree = renderer.create(<Floating {...props} />);
    expect(tree).toMatchSnapshot();
  });
});
