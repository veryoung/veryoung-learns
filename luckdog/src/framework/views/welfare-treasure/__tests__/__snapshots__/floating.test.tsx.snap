// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`render floating should render floating success 1`] = `null`;
